File: modelers/matlab/core/DiscreteFactor.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   Copyright 2012 Analog Devices, Inc.
%
%   Licensed under the Apache License, Version 2.0 (the "License");
%   you may not use this file except in compliance with the License.
%   You may obtain a copy of the License at
%
%       http://www.apache.org/licenses/LICENSE-2.0
%
%   Unless required by applicable law or agreed to in writing, software
%   distributed under the License is distributed on an "AS IS" BASIS,
%   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%   See the License for the specific language governing permissions and
%   limitations under the License.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

classdef DiscreteFactor < Factor
    
    properties
        FactorTable;
        Belief;
        FullBelief;
        Indices;
        Domain;
    end
    
    methods
        function obj = DiscreteFactor(vectorObject,indices)
            if nargin < 2
                indices = 0:(vectorObject.size()-1);
            end
            obj@Factor(vectorObject,indices);
        end
        
        function [retval] = get.FactorTable(obj)
            table = obj.VectorObject.getFactorTable();
            retval = FactorTable(table);
        end
        
        function beliefs = get.Belief(obj)
            beliefs = obj.VectorObject.getDiscreteBeliefs(obj.VectorIndices);
            beliefs = MatrixObject.unpack(beliefs,obj.VectorIndices);
        end
        function indices = get.Indices(obj)
            var = obj.getSingleNode();
            indices = var.getPossibleBeliefIndices(0)+1;
        end
        function d = get.Domain(obj)
            indices = obj.Indices;
            vars = obj.VectorObject.getVariables();
            
            d = cell(size(indices));
            
            for r = 1:size(indices,1)
                for c = 1:size(indices,2)
                    domain = cell(vars.getSlice(c-1).getDomain().getElements());
                    d{r,c} = domain{indices(r,c)};
                end
            end
            
        end
        
        function fullBelief = get.FullBelief(obj)
            
            vars = cell(obj.VectorObject.getVariables());
            dims = zeros(1,numel(vars));
            for i = 1:length(vars)
                v = vars{i};
                dims(i) = numel(v.getDomain());
            end
            
            x = zeros(dims);
            
            beliefs = obj.Belief;
            
            for i = 1:size(obj.Indices,1)
                %Is there a better way?
                tmp = cell(1,size(obj.Indices,2));
                for j = 1:size(obj.Indices,2)
                    tmp{j} = obj.Indices(i,j);
                end
                x(tmp{:}) = beliefs(i);
            end
            
            if nargin > 1
                x = subsref(x,subs);
            end
            
            fullBelief=x;
            
        end
        
    end
    methods (Access = protected)
        
        function retval = createObject(obj,vectorObject,indices)
            retval = DiscreteFactor(vectorObject,indices);
        end
    end
    
    
end

File: solvers/java/src/main/java/com/analog/lyric/dimple/model/DiscreteFactor.java
/*******************************************************************************
*   Copyright 2012 Analog Devices, Inc.
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*       http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
********************************************************************************/

package com.analog.lyric.dimple.model;

import java.util.ArrayList;

import com.analog.lyric.dimple.factorfunctions.core.FactorFunction;
import com.analog.lyric.dimple.factorfunctions.core.IFactorTable;
import com.analog.lyric.dimple.factorfunctions.core.TableFactorFunction;

public class DiscreteFactor extends Factor
{
	/*--------------
	 * Construction
	 */
	
	public DiscreteFactor(int id, FactorFunction factorFunc,
			VariableBase[] variables)
			{
		super(id, factorFunc, variables);
		
		if (!isDiscrete())
			throw new DimpleException("ack");
	}

	/*--------------
	 * Node methods
	 */
	
	@Override
	public Discrete getConnectedNodeFlat(int i)
	{
		return (Discrete)super.getConnectedNodeFlat(i);
	}
	
	/*----------------
	 * Factor methods
	 */
	
	@Override
	public IFactorTable getFactorTable()
	{
		return getFactorFunction().getFactorTable(getDomains());
	}

	
	public int[][] getPossibleBeliefIndices()
	{
		return _solverFactor.getPossibleBeliefIndices();
	}
	

	@Override
	public void replaceVariablesWithJoint(VariableBase [] variablesToJoin, VariableBase newJoint)
	{
		//Support a mixture of variables referred to in this factor and previously not referred to in this factor
		ArrayList<INode> ports = getSiblings();
		ArrayList<VariableBase> newVariables = new ArrayList<VariableBase>();
		

		//First we figure out which variables are not currently referred to in this factor.
		for (int i = 0; i < variablesToJoin.length; i++)
		{
			boolean exists = false;
			for (int j = 0; j < ports.size(); j++)
				if (getConnectedNodeFlat(j).equals(variablesToJoin[i]))
				{
					exists = true;
					break;
				}
			
			if (!exists)
			{
				newVariables.add(variablesToJoin[i]);
			}
		}

		//Next we figure out the domain lengths of all the new variables
		DiscreteDomain [] newDomains = new DiscreteDomain[newVariables.size()];
		for (int i = 0; i < newDomains.length; i++)
			newDomains[i] = ((Discrete)newVariables.get(i)).getDiscreteDomain();
		 
		//Now, we modify the combo table to include the new variables.
		if (newDomains.length > 0)
		{
			//getFactorFunction();
			IFactorTable newTable = getFactorTable().createTableWithNewVariables(newDomains);
			setFactorFunction(new TableFactorFunction(getFactorFunction().getName(), newTable));
			
			for (VariableBase v : newVariables)
				addVariable(v);
		}

		//Now get the indices of all the variables
		int [] factorVarIndices = new int[variablesToJoin.length];
		int [] indexToJointIndex = new int[variablesToJoin.length];
		
		//Figure out which are the new variables and store a mapping
		int index = 0;
		for (int i = 0; i < ports.size(); i++)
		{
			for (int j = 0; j < variablesToJoin.length; j++)
			{
				if (getConnectedNodeFlat(i).equals(variablesToJoin[j]))
				{
					factorVarIndices[index] = i;
					indexToJointIndex[j] = index;
					index++;
					break;
				}
			}
		}

		
		//Get all the domain lengths
		DiscreteDomain [] allDomains = new DiscreteDomain[ports.size()];
		for (int i = 0; i < allDomains.length; i++)
			allDomains[i] = getConnectedNodeFlat(i).getDiscreteDomain();
		
		//Create the new combo table
		IFactorTable newTable2 =  getFactorTable().joinVariablesAndCreateNewTable(
				factorVarIndices,
				indexToJointIndex,
				allDomains,
				((Discrete)newJoint).getDiscreteDomain());
		setFactorFunction(new TableFactorFunction(getFactorFunction().getName(),newTable2));
		
		//Remove old ports.
		for (int i = 0; i < factorVarIndices.length; i++)
		{
			index = factorVarIndices[factorVarIndices.length-1-i];
			ports.remove(index);
		}
		
		//Add the new joint variable
		addVariable(newJoint);

		//Tell all old variables to remove this factor graph.
		for (VariableBase v : variablesToJoin)
			v.remove(this);
		
		
	}

	public String getFactorTableString()
	{
		String s = "TableFactor [" + getLabel() + "] " + getFactorTable().toString();
		return s;
	}
	
	public double [] getBelief()
	{
		return (double[])_solverFactor.getBelief();
	}


}
